### PR TITLE
feat: implement PhoneNumberInput component for consistent phone numbe…

### PR DESCRIPTION
**Pull request by farmanahmed888**
This pull request introduces a unified approach for handling Indian mobile phone numbers across both the Content Web App and Teacher Web App. It adds reusable `PhoneNumberInput` components, centralizes phone number validation and normalization logic, and updates forms and API interactions to ensure consistent, sanitized, and validated phone numbers. Comprehensive tests are included for the new utilities and components.

**Phone number input and validation improvements:**

* Added a reusable `PhoneNumberInput` component in both apps (`ContentWebApp/src/components/AllContent/shared/PhoneNumberInput.js`, `teacher-webapp/src/components/common/PhoneNumberInput.js`) that restricts input to digits only and a maximum of 10 digits, ensuring consistent user experience and validation. [[1]](diffhunk://#diff-55ba695c1f26b74925d0716e5333ca86c7dde4ba066f3ed04281fef6abe89581R1-R36) [[2]](diffhunk://#diff-c16238d4bd7735656a22305b5467170217ce7bd74353392154eb2b22d8d5a3fbR1-R51)
* Updated registration and student-add forms to use the new `PhoneNumberInput` and enforce 10-digit validation with user feedback for errors. [[1]](diffhunk://#diff-89b5f71eb4931aaaf52797996b0d28e6937221ead42f5c953c69c5884309f6beR5) [[2]](diffhunk://#diff-89b5f71eb4931aaaf52797996b0d28e6937221ead42f5c953c69c5884309f6beL24-R29) [[3]](diffhunk://#diff-6fd0a583f56e3fad06c1efcb78db0ba545432759844f63414de3cdf275b65394R6-R19) [[4]](diffhunk://#diff-6fd0a583f56e3fad06c1efcb78db0ba545432759844f63414de3cdf275b65394L25-R37) [[5]](diffhunk://#diff-6fd0a583f56e3fad06c1efcb78db0ba545432759844f63414de3cdf275b65394L53-R58) [[6]](diffhunk://#diff-774dfede8c6eb7b7c8d25ec1ab6fc77e0520ef252ca2f319abaeea78ef67eaeaR37-R42) [[7]](diffhunk://#diff-1bf2103dabf8ad4b95c3ff091cfa2d0b5c9a2e17a58032b12f7cd03e71565957L15-R23) [[8]](diffhunk://#diff-1bf2103dabf8ad4b95c3ff091cfa2d0b5c9a2e17a58032b12f7cd03e71565957R72-R82)

**Centralized phone number utilities:**

* Added `phoneUtils.js` with functions for sanitizing, normalizing, and formatting phone numbers for API use, including `normalizeToTwelveDigits`, `normalizePhoneNumber`, and `sanitizePhoneInput`.
* Updated all API calls and data processing (registration, adding/removing students, classroom forms) to use these utilities, ensuring only valid, normalized phone numbers are sent to the backend. [[1]](diffhunk://#diff-e6df0fa093ee0ba38650ea097cde7960b34907952eaed0ed733b6949f366a9f1R4) [[2]](diffhunk://#diff-e6df0fa093ee0ba38650ea097cde7960b34907952eaed0ed733b6949f366a9f1L68-R76) [[3]](diffhunk://#diff-e6df0fa093ee0ba38650ea097cde7960b34907952eaed0ed733b6949f366a9f1L152-R190) [[4]](diffhunk://#diff-e6df0fa093ee0ba38650ea097cde7960b34907952eaed0ed733b6949f366a9f1R219-R244) [[5]](diffhunk://#diff-d5c96045d2a02d130c3b6919fe4c1e5e2253efdec79193b3f6432190ad213791R35) [[6]](diffhunk://#diff-d5c96045d2a02d130c3b6919fe4c1e5e2253efdec79193b3f6432190ad213791R195-R205)

**Testing and reliability:**

* Added comprehensive unit tests for the phone number utility functions and the new `PhoneNumberInput` component to ensure correct behavior and prevent regressions. [[1]](diffhunk://#diff-3e011928f5ff9523c0d4efa1e552459418ba227d147b7b5855e4a47bed90cc2bR1-R88) [[2]](diffhunk://#diff-9b7f72f4b267bafb3368bc5c906a683240fac1475136aaf74c8ab87fbdde52cbR1-R36)

**Other minor cleanup:**

* Removed unused imports and improved error handling and messaging for invalid phone numbers in various flows. [[1]](diffhunk://#diff-3157ebb9fc36a49a9275ff7bf9ac4028dd3d9fd922ac03dfac5b4a049ab98334L12) [[2]](diffhunk://#diff-e6df0fa093ee0ba38650ea097cde7960b34907952eaed0ed733b6949f366a9f1R219-R244)

This refactor ensures that phone numbers are consistently validated, sanitized, and formatted throughout the applications, reducing bugs and improving user experience.

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
**Summary of Review:**

The main focus is on consistent, clean handling of phone number normalization across components and hooks. The code now introduces a dedicated `PhoneNumberInput` component with proper input sanitation and length restrictions. Utility functions are structured correctly to handle normalization (e.g., `normalizeToTwelveDigits`) before API calls, reducing repeated logic.

**Key issues & recommendations:**

- **Type Definitions & Consistency:** Some functions like `normalizePhoneNumber` have ambiguous comments; clarify input types (string vs. number) and ensure they return consistent formats.
- **Redundant/Over-complicated Logic:** 
  - `normalizePhoneNumber` and `normalizeToTwelveDigits` overlap in functionality; consider consolidating or clearly delineating their purposes.
- **Neglected edge cases:** Ensure all normalization functions handle invalid or unexpected inputs gracefully.
- **Testing:** Tests are comprehensive but could include edge cases (non-string inputs, prefix variations).
- **Design Pattern:** Proper separation of UI (input component) and normalization (utility functions) is good. However, `normalizeForApi` might be unnecessary if directly wrapping `normalizePhoneNumber`.
- **Cyclomatic Complexity:** Not excessive; logic is straightforward.

**Overall Score:** 85/100
- The code is improved, maintenance is easier, but minor polishing and clearer documentation needed.

**Next Steps:**  
- Consolidate normalization functions.  
- Enforce strict input types and validation.  
- Keep testing edge cases.  
- Maintain clear separation of concerns.

**Acceptance with suggestions for minor refinement.**

**Review by theainerd — APPROVED** [review]
**Verdict**: Approved

Excellent work on the `PhoneNumberInput` component and the robust normalization logic in `phoneUtils.js`. The test coverage gives confidence in the changes.

LGTM!

**Review by a4i-architect — COMMENTED** [review]
### High-Level Summary & Review

The changes mostly implement a consistent approach for handling Indian phone numbers across various components and utils by normalizing input and API formats. The introduction of `PhoneNumberInput` components improves input validation and UX.

### Key Improvements & Issues
- **Redundancy & Duplication**: Several components have similar validation logic, especially regarding phone inputs. Consider centralizing phone validation and normalization logic.
- **Type Handling & Robustness**: In `phoneUtils.js`, functions like `normalizePhoneNumber` do not handle non-string inputs robustly (e.g., null, numbers). Wrap with type checks.
- **Inconsistent Validations**: `normalizeToTwelveDigits` only checks regex; further validation might be needed before API call to prevent errors.
- **Unnecessary Comments/Misleading Warnings**: Some warnings in utils are inconsistent since `normalizePhoneNumber` may be called with already sanitized inputs.
- **Design & Pattern**: Good abstraction with `PhoneNumberInput`; however, the multiple `normalize` functions suggest a need for a unified normalization service.
- **Null/Undefined/Edge Cases**: Tests and code should explicitly handle null/undefined inputs to prevent runtime errors.

### Scoring & Conclusion
- *Sufficient improvements in phone input handling, but inconsistent validation/robustness lowers quality*.
- **Score: 75/100** (acceptable with internal improvements).

### Recommended Actions
- Consolidate phone normalization in utils with strict type checks.
- Remove redundant validation between components and utils.
- Enhance handling of edge cases and invalid inputs.
- Consider creating a centralized phone validation/formatting service.

**References:**  
- [Clean Code Patterns](https://refactoring.guru/design-patterns)  
- [Handling Phone Numbers in Frontend](https://developers.google.com/maps/documentation/geocoding/overview#rangecodes)

Proceed with acceptance after addressing robustness concerns.

**Review by a4i-architect — COMMENTED** [review]
### Summary of Review

The code moves to a more unified, sanitizing approach for handling phone inputs with the new `PhoneNumberInput` components, replacing raw `<input>` fields across multiple forms. This improves data consistency and validation.

### Issues & Recommendations

1. **Code Duplication & Variability**:
   - Two distinct `PhoneNumberInput` components exist: one from `/shared/PhoneNumberInput.js` (React, simple) and another from `/components/common/PhoneNumberInput.js` (Material UI, more advanced). Consider consolidating to a single, configurable component to avoid maintenance overhead and confusion.
   
2. **Type and Validation Handling**:
   - `onChange` handlers now primarily pass sanitized strings. Ensure parent components validate length explicitly (e.g., 10 digits) before API use.
   - For `normalizeToTwelveDigits`, check if it should return null or empty string for invalid input; current logic is okay, but enforce strict validation before API calls.

3. **Redundant or Missing Checks**:
   - `useTeachers.js` code snippet has a `setMessage` call on error, but `setMessage` isn't defined or imported; likely a bug or incomplete. Consider adding proper error state management.

4. **UI/UX Consistency**:
   - The custom `<PhoneNumberInput>` components uniformly restrict input length and sanitize onChange, which is good. Ensure accessibility is maintained (e.g., aria labels).

5. **Design & Complexity**:
   - The `phoneUtils.js` functions are mostly straightforward; avoid overly complex normalization logic to reduce cyclomatic complexity.
   - Consider removing duplicate utils or consolidating methods dealing with normalized formats.

### Score & Final Decision

**Score: 75/100**

- The migration to sanitized phone input components enhances data hygiene.
- Minor issues with incomplete code snippets, redundant components, and potential error handling bugs need fixing.
- Overall good, but slight refactoring and cleanup are recommended before acceptance.

**Next steps**:
- Consolidate `PhoneNumberInput` components.
- Fix missing error state (`setMessage`) and ensure parent forms validate length before API calls.
- Confirm accessibility and validation consistency.

**References**:
- Effective Input Sanitization: [React Best Practices](https://reactjs.org/docs/forms.html#handling-multiple-inputs)
- Utility function design: [Writing Pure Utility Functions](https://kentcdodds.com/blog/how-to-write-great-js-utilities)

**Review by a4i-architect — COMMENTED** [review]
### High-Level Review Summary

Key positive:  
- Consistent use of a custom `PhoneNumberInput` component for phone fields improves input control.
- Introduction of `phoneUtils.js` functions (`sanitizePhoneInput`, `isValidPhoneNumber`) standardizes phone validation.
- Improved UX feedback: error messaging when invalid phone numbers are entered before submission.

Issues & Recommendations:  
1. **Redundant and Overlapping Code**:
   - Multiple `PhoneNumberInput` implementations (`shared` and `common`) exist, leading to inconsistent UI/UX.
   - Focus on consolidating to one, well-tested version (preferably the Material-UI version for consistency).

2. **`phoneUtils.js` Design**:
   - The normalization functions (`normalizePhoneNumber`) are inconsistent; in some places, they suggest prefixing ("91") which isn't reflected in actual code.  
   - Clarify if backend expects 10-digit or 12-digit ("91" prefix). Currently, the code transmits 10 digits only, which is acceptable if backend expects that.

3. **Type & Validation Logic**:
   - Be consistent: `isValidPhoneNumber` correctly strips non-digits, but ensure all call sites uniformly validate corresponding inputs.
   - In hooks (`useTeachers.js`), validation now uses `isValidPhoneNumber`. It's good, but ensure all related logic (e.g., form submission) aligns.

4. **Code Quality & Best Practices**:
   - Avoid duplicating validation logic; centralize in `phoneUtils.js`.
   - Remove commented or dead code snippets to improve clarity.
   - Use React's strict types or PropTypes for component props.

5. **Testing**:
   - Tests are comprehensive for `PhoneNumberInput` and `phoneUtils.js`, which is good.
   - Ensure all UI inputs are covered (e.g., parent forms interest in controlled/disabled states).

### Design & Complexity
- The use of `sanitizePhoneInput` and `isValidPhoneNumber` is a good pattern, reducing code duplication.
- Complexity is low; no over-complicated logic.
- Consolidating on a single `PhoneNumberInput` component and unified validation will reduce cyclomatic complexity and improve maintainability.

### Final Assessment
- The approach is solid, but needs consolidation of phone input variations, cleaning of redundant code, and clear validation flow.
- Score: **78/100**  
  (Acceptable after minor refactoring and cleanup)

### Next Steps
- Standardize on one `PhoneNumberInput` component (prefer Material-UI or shared, but not both).
- Remove unused or duplicate code.
- Verify backend expects 10-digit; if not, update normalization.
- Maintain consistent validation everywhere.
- Keep test coverage comprehensive.

**Reference Articles**:  
- [React Controlled vs Uncontrolled Inputs](https://reactjs.org/docs/uncontrolled-components.html)  
- [Centralizing Validation Logic](https://uxdesign.cc/best-practices-to-validate-forms-in-react-84d4b56993e7)

File: ContentWebApp/src/hooks/useTeachers.js
Comment (N1kunj1998) [inline ContentWebApp/src/hooks/useTeachers.js:153]: Here there is a bug, user can by mistake enter phone number lenght of 8 or 9 when trying to add multiple students in that case instead of dropping invalid phone numbers we can give user message that one of the phone number is incomplete.

```
      const mappedStudents = (teacher.newStudents || []).map((s) => ({
        name: (s.name || "").trim(),
        phoneNumber: (s.phoneNumber || "").trim(),
      }));

      const filledRows = mappedStudents.filter((s) => s.name || s.phoneNumber);
      const invalidRows = filledRows.filter(
        (s) => s.name && s.phoneNumber && !isValidPhoneNumber(s.phoneNumber)
      );

      if (invalidRows.length > 0) {
        setMessage(
          `${invalidRows.length} row(s) have incomplete phone numbers. Please enter exactly 10 digits.`
        );
        setTimeout(() => setMessage(""), 4000);
        return;
      }

      const payloadStudents = mappedStudents.filter(
        (s) => s.name && s.phoneNumber && isValidPhoneNumber(s.phoneNumber)
      );
```

File: ContentWebApp/src/components/AllContent/shared/PhoneNumberInput.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  autoFocus = false,
+  readOnly = false,
+}) => {
+  const handleChange = (e) => {
+    onChange(sanitizePhoneInput(e.target.value));
+  };
+
+  return (
+    <input
+      type="tel"
```
Comment (N1kunj1998) [inline ContentWebApp/src/components/AllContent/shared/PhoneNumberInput.js:24]: we should have here required prop, just like we are having in teacher-webapp

File: ContentWebApp/src/hooks/useTeachers.js
Comment (farmanahmed888) [reply ContentWebApp/src/hooks/useTeachers.js:153]: image

we already do this without sending a request to backend see line number 158 in  `useTeachers.js`

``` js
.filter((s) => s.name && s.phoneNumber && isValidPhoneNumber(s.phoneNumber))
```

File: ContentWebApp/src/components/AllContent/shared/PhoneNumberInput.js (hunk 1 above)
Comment (farmanahmed888) [reply ContentWebApp/src/components/AllContent/shared/PhoneNumberInput.js:24]: done

**Review by a4i-architect — COMMENTED** [review]
**Summary & Key Issues:**

- **Consistent Phone Input Handling:**  
  Multiple components (`AddStudentsForm`, `TeacherRegistrationForm`, `Login`, `Register`, `common/PhoneNumberInput`) now use a custom `PhoneNumberInput` with `onChange` accepting sanitized or plain values. However:
  - The parent components differ: some use `value` directly, others manage separate state. Ensure all handle sanitized values and enforce `maxLength` uniformly.
  - In `TeacherRegistrationForm`, `onChange={setTeacherPhone}` passes a string, which is correct. But in `AddStudentsForm`, the handler uses `e.target.value`, possibly un sanitzed. Confirm all handlers sanitize reliably.
  
- **Redundant/Overlapping Validation:**  
  Validation functions (`isValidPhoneNumber`) are now used in various places (hooks, forms). This centralizes validation but ensure all handle raw input data correctly. Some components check length explicitly; prefer using validation functions everywhere for consistency.

- **CSS/UX Improvements:**
  - Consistent error handling: in `TeacherRegistrationForm`, a dedicated error message style added; other forms should adopt similar.
  - The `PhoneNumberInput` in `teacher-webapp/src/components/common` is more sophisticated, using Material-UI, which is good. Ensure design consistency across app.

- **Code Duplication Concerns:**  
  Multiple similar utility functions (`sanitizePhoneInput`, `isValidPhoneNumber`) are well abstracted. No major duplication.

- **Type and Logic Checks:**  
  Functions are type-safe; ensure all handlers pass sanitized string values, avoiding unexpected behaviors.

- **Component Usage & Interface:**  
  Using custom input components improves UX; however, verify that all parent components are compatible with expected `value` and `onChange` props.

**Design & Complexity:**

- Overall, a move toward consistent, sanitized input handling with validation functions is positive.
- No complex cyclomatic issues observed; prefer centralized validation and uniform API for custom inputs.

---

### Recommendations:

- Enforce consistent `onChange` signatures—preferably always passing sanitized values.
- Centralize all phone validation logic in `phoneUtils`.
- Review all form handlers to prevent passing raw or unvalidated data.
- Add error states for all forms with styled feedback.
- Consider removing duplicate validation logic in favor of only using `isValidPhoneNumber`.

### Score: **78/100**  
*(Improvements needed for uniformity and code hygiene, though the new input handling is solid.)*

**Review by N1kunj1998 — APPROVED** [review]
Everything working fine on local

**Review by a4i-architect — COMMENTED** [review]
### Summary and Key Findings

**Positive Aspects:**
- Introduction of a custom `PhoneNumberInput` component for consistent phone input handling.
- Proper client-side validation for phone number length (10 digits).
- Code coverage via tests for `phoneUtils` and `PhoneNumberInput`.

**Issues & Recommendations:**

1. **Redundant/Overlapping Components:**
   - Two separate `PhoneNumberInput` components exist (`../shared/PhoneNumberInput.js` and `common/PhoneNumberInput.js`) with different styles and APIs. This duplication increases complexity and maintenance overhead.
   - Standardize on a single, unified `PhoneNumberInput` component with a well-defined API.

2. **Validation & Error Handling:**
   - Validation logic (e.g., `isValidPhoneNumber`) is duplicated across the utils, hooks, and form components, which is good. However, some places (like login) still manually check input length, leading to inconsistent UX.
   - Ensure all forms use the centralized validation utility before submission.

3. **Type Safety & Robustness:**
   - The utils functions (`sanitizePhoneInput`, `isValidPhoneNumber`) are reasonably safe but could benefit from explicit type annotations (TypeScript) or runtime type checks.
   - Some backend assumptions (e.g., always 10-digit valid numbers) are enforced but could be enhanced with more user feedback (e.g., real-time validation).

4. **Code Complexity & Design:**
   - The `useTeachers` hook contains some complex, repeated useCallback hooks, which could be simplified for clarity, especially the dependency arrays.
   - The form validation logic for students and leaders is somewhat verbose; consider extracting validation functions for clarity.
   - The `formatStudentPhones` function is appropriately designed.

5. **Testing:**
   - Tests for `PhoneNumberInput` are good, but some component imports have inconsistent structure/UI and should match the final component API.
   - Consider adding tests for `isValidPhoneNumber` edge cases and for the entire validation flow.

6. **Miscellaneous:**
   - The `teacher-webapp` files show duplicated or inconsistent validation logic for phone number inputs across various components.
   - The use of inline validation and manual sanitization could be consolidated into a single, reusable utility.

### Overall Score: 78/100
- The foundation is solid, with good validation practices.
- The major flaw is component duplication and potential confusion from inconsistent API design.
- With refactoring to unify `PhoneNumberInput`, cleanup of hooks, and enhanced validation UX, score could improve.

**Acceptance is conditional** upon:
- Consolidating `PhoneNumberInput` components.
- Improving validation consistency.
- Simplifying hooks and form logic for maintainability.

**References:**
- [React Custom Inputs & Validation](https://reactjs.org/docs/forms.html)
- [Design Patterns for Form Validation](https://uxdesign.cc/form-validation-best-practices-1c8a2b9aad0f)
- [Clean Code & React Hooks Best Practices](https://reactjs.org/docs/hooks-overview.html)

**Review by a4i-architect — COMMENTED** [review]
### High-Level Summary
The codebase integrates a custom `PhoneNumberInput` component throughout registration and form pages, enforcing 10-digit input validation, with utility functions for sanitization and validation. Tests have been added to verify input sanitization and validation behavior.

### Issues & Recommendations
1. **Code Duplication & Inconsistent Components**: There are multiple `PhoneNumberInput` components with different implementations (e.g., one from shared components, another from common). Consolidate into a single, stable, well-documented version to avoid confusion.
2. **Type & Validation Logic**: Use `isValidPhoneNumber()` consistently. Currently, some places (like `useTeachers.js`) validate the length directly, which is redundant given the utility functions.
3. **Input Handling**: The `PhoneNumberInput` components correctly sanitize on change, but some parent components still perform length checks—prefer centralizing validation via utility functions.
4. **Testing**: Tests are comprehensive for sanitization/validation but should also cover edge cases like empty input, null, or special characters explicitly.
5. **Design & Complexity**: The code structure is straightforward. No over-complication detected.

### Score: 80/100
**Accept** with action items:
- Standardize the `PhoneNumberInput` component (prefer the one under `components/common/`) across the app.
- Ensure all validation uniformly relies on `isValidPhoneNumber()`.
- Extend tests to cover all edge cases.
- Remove duplicate utility functions or components.

**References**:
- [React Controlled Components](https://reactjs.org/docs/forms.html)
- [Input Validation Best Practices](https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation)

File: teacher-webapp/src/pages/Login.js [hunk 2]
```diff
@@ -123,32 +130,11 @@ function Login() {
...
-                startAdornment: (
-                  <InputAdornment position="start">
-                    <PhoneIcon />
-                  </InputAdornment>
-                ),
-              }}
-              aria-label="Phone number input"
-              aria-required="true"
+              onChange={setPhoneNumber}
               onKeyPress={handleKeyPress}
```
Comment (theainerd) [inline teacher-webapp/src/pages/Login.js]: `onKeyPress` is deprecated in React's synthetic event system and may not be reliably forwarded by MUI's `TextField` to the native `<input>` element across all versions. This could cause the Enter-to-submit behavior to silently stop working. Consider migrating to `onKeyDown` for a more robust and future-proof approach.

**Review by theainerd — COMMENTED** [review]
Overall the validation and sanitization approach is solid — centralizing phone number logic into reusable components and utilities is a clear improvement. A couple of items worth addressing below.

File: teacher-webapp/src/pages/ClassroomForm.js [hunk 3]
```diff
@@ -59,9 +59,9 @@ const ClassroomForm = () => {
     if (isEditMode) {
       fetchClassroom();
     }
-  }, [classroomId]);
+  }, [classroomId, isEditMode, fetchClassroom, fetchTeacherStudents]);
```
Comment (theainerd) [inline teacher-webapp/src/pages/ClassroomForm.js]: The `useEffect` at line 57 references `fetchClassroom` and `fetchTeacherStudents` in its dependency array, but both are `const` declarations (via `useCallback`) defined after it (lines 64 and 87). While React's hook execution order prevents a runtime crash in practice, `const` bindings are in the Temporal Dead Zone before their declaration, which makes this pattern fragile and harder to reason about. Moving the `useCallback` definitions above this `useEffect` would make the dependency flow clearer.

File: teacher-webapp/src/pages/Login.js (hunk 2 above)
Comment (farmanahmed888) [reply teacher-webapp/src/pages/Login.js]: done

File: teacher-webapp/src/pages/ClassroomForm.js (hunk 3 above)
Comment (farmanahmed888) [reply teacher-webapp/src/pages/ClassroomForm.js]: done

**Review by a4i-architect — COMMENTED** [review]
### Summary

The review introduces a custom `PhoneNumberInput` component to handle digit-only, max 10-digit phone input with sanitization and validation. This enhances input consistency across forms. The component improves UX by restricting input length and sanitizing values on change, avoiding manual pattern handling.

### Strengths
- Consistent phone input handling via a dedicated component.
- Validates phone number length before submitting forms.
- Tests cover sanitization and validation thoroughly.

### Issues & Recommendations
- **Redundant code & type handling**: Multiple instances of `onChange` handlers re-sanitizing input; unify with the new component.
- **Type clarity**: The custom input's `onChange` expects a string (sanitized value), but consuming components sometimes funnel raw event objects. Enforce explicit prop expectations and typings.
- **Validation logic**: Repetitive length checks; centralize validation (e.g., `isValidPhoneNumber`) to avoid duplication.
- **Input component consistency**: Some forms (e.g., Login, Register) still manually strip digits; switch fully to `PhoneNumberInput` for uniformity.
- **Testing improvements**: Tests are good but could add edge cases (empty input, partial input).

### Design & Complexity
- The core design of wrapping `input` in a sanitized, length-limited component is solid.
- Keep validation logic single-source to reduce potential bugs.
- The code complexity remains low; no problematic cyclomatic growth observed.

### Final score: **82/100**

**Acceptable after:**
- Fully replacing manual phone handling in forms with `PhoneNumberInput`.
- Centralizing validation logic.
- Clarifying prop types/documentation, especially for `onChange`.

**References:**
- [Custom Input Components in React](https://reactjs.org/docs/forms.html)
- [Validation best practices](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/HTML5/Validation)
- [Handling Input Sanitization](https://medium.com/@ryanbedford/sanitizing-and-validating-user-input-in-react-2a3193571772)
